packages/dependency-extraction-webpack-plugin/lib/index.js

```diff
@@ -198,14 +198,25 @@ class DependencyExtractionWebpackPlugin {
 				}
 			}
 
+			// Go through the assets and hash the sources. We can't just use
+			// `entrypointChunk.contentHash` because that's not updated when
+			// assets are minified. Sigh.
+			// @todo Use `asset.info.contenthash` if we can make sure it's reliably set.
```

Comment:
No, that's not what it means.

What this comment is referring to is that _sometimes_ Webpack sets `.info.contenthash` on the `asset` object, but it only does so if it decides that it needs it (if I recall correctly, it depends on whether the filename template uses `[contenthash]`). If we could ensure that Webpack sets that every time, we could just use it instead of hashing the file contents ourself.

Reply:
Got it. I think the algorithm you proposed is a nice improvement so let's rephrase the comment so it presents the benefits, instead of the limitations of the webpack. I can fully confirm that it's going to solve the issue that Jetpack struggles with as explained by @kraftbj with hashes changing depending on the absolute path for files. We can also link this PR so people can learn the full context.

Reply:
I have no objection to changing the comment, but I don't know what to rephrase it to. Care to make a suggestion?

Reply:
If I understand it correctly, ideally we would just change (line 208):
```
version: entrypointChunk.hash
```
to:
```
version: entrypointChunk.contenthash
```
and be done? But we can't do that because webpack doesn't always set it? And/or the value doesn't have the properties that we need?

I noticed that webpack has an [`optimization.realContentHash`](https://webpack.js.org/configuration/optimization/#optimizationrealcontenthash) config option. Which means that there are multiple valid ways how to compute `contenthash`, and that the default ways is in some sense "not real."

Reply:
> But we can't do that because webpack doesn't always set it? 

That one we can't use because it doesn't get updated when the minifier changes the code. It's `asset.info.contentHash` that isn't always set.

Reply:
> That one we can't use because it doesn't get updated when the minifier changes the code.

I tested and debugged this a little and found something interesting. Chunk filenames are generated using the `chunk.contentHash` value, which is calculated from the unminified and unoptimized code. Not great. But then, one of the last asset processing stages is `PROCESS_ASSETS_STAGE_OPTIMIZE_HASH`, where the `RealContentHashPlugin` (enabled by default) does its work, and it updates the hashes, and renames the assets. These updated hashes are calculated from the minified code.

Our plugin runs in the `PROCESS_ASSETS_STAGE_ANALYSE`, which is after the hashes got updated. So in theory we should have access to them. But apparently they are calculated and updated only when the file name actually used the `[contenthash]` token? Or they are calculated always and we just haven't figured how to get them?

We don't use `[contenthash]` in file names, but we use it to generate a special `.php` asset which is then used by the WordPress script enqueueing engine to produce a `filename.js?ver=contenthash` script URL.

The one bad thing about this PR is that we're calculating the content hash ourselves, instead of letting webpack do it. I wish there was a way to fix that. Maybe there is really a webpack bug -- wanting to use the updated contenthash although it's not used in filenames is a reasonable demand.

Reply:
> But apparently they are calculated and updated only when the file name actually used the `[contenthash]` token?

Yes, exactly. Which is why we can't take advantage of it being set here.

> The one bad thing about this PR is that we're calculating the content hash ourselves, instead of letting webpack do it.

Thus the `@todo` comment here. I hope we can avoid letting the perfect be the enemy of the good, and get this merged even though we have to calculate the hash outselves instead of trying to get Webpack to first develop a new feature.

Reply:
If we have the goal to use the same hash as the real `[contenthash]`, and we are only forced to calculate it ourselves, they should match. I.e., if my webpack config uses a `[name].[contenthash].js` filename template, and also uses the extraction plugin, the `[contenthash]` and `version` should be the same. But they currently aren't. The differences I see are:
1) webpack uses `md4` algorithm by default, configurable with `output.hashFunction` option, while the plugin uses `sha512`.
2) webpack uses only the assets' contents to update the hash, while we hash the filename, too.
It's weird that even making these two changes, the hashes are still different for me.

It would be worthwhile to spend some timeboxed time trying to fix this before merging. Other than that, I think this patch is ready to land.

Reply:
`md4` no longer works with Node 17+.

Replicating the same algorithm as webpack uses internally is rather complex. This is the initial commit when the support for `contentHash` was added:
https://github.com/webpack/webpack/commit/b929d4cdce433ac91fd8e4e5b8e6c8e90c6d5ca0.

Maybe we should try to detect `contentHash` first, and fall back to the custom handler otherwise. Example `contentHash` objects:

```js
{
  'css/mini-extract': '9dc6bf4629f53268df7c',
  javascript: 'bc28cb02479bf7449a77'
}
{ javascript: '684ed11ffa88cd017286' }
```

Reply:
> `md4` no longer works with Node 17+.

webpack ships its own WASM module that implements `md4`. Instead of importing `createHash` from Node's `crypto` module, we can use the webpack version:
```js
const webpack = require( 'webpack' );
const hash = webpack.util.createHash( 'md4' );
```

> Replicating the same algorithm as webpack uses internally is rather complex.

The `RealContentHashPlugin` source is indeed very complex, and I'm not sure what set of possible use cases it covers, but I've been able to replicate the `contenthash` quite easily:
```js
const hash = createHash( 'md4' );
// asset.source.updateHash( hash );
hash.update( asset.source.buffer() );
const version = hash.digest( 'hex' );
```
Here the non-obvious step is to not use `asset.source.updateHash`, because that hashes a string `"RawSource" + source`, instead of just `source`. That's all the difference and when computing a content hash, we only want the source.

Reply:
> Here the non-obvious step is to not use `asset.source.updateHash`, because that hashes a string `"RawSource" + source`, instead of just `source`. That's all the difference and when computing a content hash, we only want the source.

Looking at it another way, we don't actually care what exactly goes into the hash as long as changes to the output result in changes to the hash, and non-changes to the output do not result in changes to the hash. Whether or not `"RawSource"` is incorporated in the hash makes no difference from that perspective.

OTOH, using `asset.source.updateHash` may have a performance advantage if the source is something like [ReplaceSource](https://github.com/webpack/webpack-sources/blob/main/lib/ReplaceSource.js) or [ConcatSource](https://github.com/webpack/webpack-sources/blob/main/lib/ConcatSource.js) that does non-trivial work in `source()` or `buffer()`.